### PR TITLE
Directly connect health check to GazetteerCache and remove the use of threading in model training

### PR DESCRIPTION
**Pull request by jwalgran**
## Overview

Directly connect health check to GazetteerCache, remove the use of threading in model training, fix related issues, and add additional request logging support.

Connects https://github.com/open-apparel-registry/open-apparel-registry-clients/issues/11
Connects #1122
Fixes https://github.com/open-apparel-registry/open-apparel-registry/issues/1091

## Notes

- We added the `postfacilitiescsv.py` command line script as a repeatable load testing utilty that takes a CSV with name, address, and country_name/country columns and POSTs each line to the facility matching API.
  - For convenience the token and root URL can be passed as flagged arguments or set as TOKEN and ROOTURL environment variables. The values default to those that will work with the development environment populated with `resetdb`.
  - The -s,--stop argument allows running tests with small numbers of requests without needing to create input CSVs of different lengths.
  - We added some diagnostic `id` and `index` query string arguments so that we could differentiate the activity of multiple simultaneous runs of the tool in request logs and determine if requests are being retired by CloudFront.
- We added a development heath check tool as an attempt to mimic, in as simple a way as possible, a repeating call to the heath check endpoint that is vaguely similar to the request an AWS ALB would make.
- Under load from facility matching API requests our app service tasks were being marked as unhealthy and being restarted. It was not clear from the logs in AWS exactly why our services were being marked as unhealthy so we have atempted to simplify the code around the `GazetteerCache` to make failures easier to reason about.
  - By adding a custom watchman check that calls `GazetteerCache.get_latest` we ensure that the task will not be marked as healthy until a gazetteer matcher has been fully trained.
  - We added an ECS service health check grace period to allow for model training time we initial tested with 60 seconds and increased to 180 seconds which was the point at which the instance would be successfully marked as healthy.
  - We set the gunicorn timeout to 120 seconds because saw `[CRITICAL] WORKER TIMEOUT` error messages in the ECS task log output when this was set to a lower value. 90 matches the timeout value we set in docker-compose.yml for development but it appears that the ECS tasks may require more time.
- We want our model training to be part of the initialization of a service tasks so we add an `AppConfig` subclass to move this process earlier in the Django "boot up" process.
  - We copied a check of the `SERVER_SOFTWARE` environment variable from the old `startup.py` which ensures that `get_latest` is only called when the app is running in a Gunicorn worker, not when the app was loaded to run a management command.
  - We are keeping the custom health check call to `get_latest` because it will help continuously keep the GazetteerCache "fresh" in cases where a list is uploaded and processed by our batch workflow.
- If each gunicorn worker is a process, it must have its own, trained gazetteer matcher. We set the workers to 1 so that each Fargate task has a single gazetter matcher.
- While testing the revised API service we encountered and fixed some related issues.
  - The Django admin was reading every `Source` in order to populate a dropdown list on the facility list item admin page. Our production database now has over 100,000 source records so loading these into a dropdown is not useful and resource intensive, in some cases causing service outages by tying up gunincorn workers for an extended period of time. There is not currently a need to ever change the source for a line item, so it is safe to make it read-only.
  - Submitting facilities through the API with create=false will create a `FacilityListItem` record but not a FacilityMatch. These records referencing a `Facility` were preventing that facility from being deleted as part of a merge opertation, so we added a block to the merge API view function to properly handle these records.
  - We were previously calling `cls._gazetter.unindex` to remove records with a `history_type` of `-` but it was raising exceptions for which we could not determine the root cause. We have opted to ignore them and filter out no longer existing records from the match results.
- We added support for querying CloudFront and Application Load Balancer logs via Athena to allow reporting and analysis of our API request load.

## Testing Instructions

This PR underwent multiple tests on staging with an external API partner and internally using `postfacilitiescsv.py` and sample data provided by the API partner.
 
## Checklist

- [x] `fixup!` commits have been squashed
- [ ] CI passes after rebase
- [x] CHANGELOG.md updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines

### CODE REVIEW

**Comment by jwalgran** [comment]
Testing notes copied from #1108 

To generate test data for exercising the API I went through the following steps:

- Downloaded the full list of facilities from openapparel.org and used the following script to extract only the data we need:
```python
#!/usr/bin/env python

import sys
import fileinput
import csv

# Read a CSV downloaded from OAR search results on standard input and
# write a 5-column CSV to standard output suitable for uploading as a
# contributed list.

# USAGE
#
# split_download < facility_list.csv > facility_list_to_upload.csv

csv_reader = csv.reader(fileinput.input(), delimiter=',')
next(csv_reader)
csv_writer = csv.writer(sys.stdout)
csv_writer.writerow(['country', 'name', 'address', 'lat', 'lng'])
for row in csv_reader:
    csv_writer.writerow([row[4], row[1], row[2], row[5], row[6]])
```
- Used following bash function (taken from https://stackoverflow.com/a/20721203) to split that file into 5000 record chunks and copied those chunks to `data-dumps-for-testing/segmented-prod-csv` on the file share. **NOTE** that this uses `gsed` which I installed on my Mac via homebrew because the `sed` on macOS works differently than GNU sed.
```bash
splitCsv() {
    HEADER=$(head -1 $1)
    if [ -n "$2" ]; then
        CHUNK=$2
    else 
        CHUNK=1000
    fi
    tail -n +2 $1 | split -l $CHUNK - $1_split_
    for i in $1_split_*; do
        gsed -i -e "1i$HEADER" "$i"
    done
}
```
- Ran `./scripts/resetdb` in development
- Logged in to development as c2@example.com uploaded the first 5000-record chunk and processed it with management commands using a bash helper function 
```bash
function batch_process {     
    ./scripts/manage batch_process --list-id $1 --action parse  && \
    ./scripts/manage batch_process --list-id $1 --action geocode  && \
    ./scripts/manage batch_process --list-id $1 --action match; 
}
```

```
batch_process 16
```

- Refreshed the list page, filtered by the status of `POTENTIAL_MATCH` and resolved all potential matches.
- Repeated the process for each chunk, logging in as a different c#@example.com user.
- Once all the chunks were processed I used `pg_dump` with no arguments to dump the dev database with ~42000 facilities to a .sql file and put it on the file share.

**Comment by jwalgran** [comment]
While running 4 streams of POST requests, the staging services did become unhealthy and were replaced. I saw 504s, a 502, and 503s appear in the responses

```
Status: 504 Content: b'<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">\n<HTML><HEAD><META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">\n<TITLE>ERROR: The request could not be satisfied</TITLE>\n</HEAD><BODY>\n<H1>504 ERROR</H1>\n<H2>The request could not be satisfied.</H2>\n<HR noshade size="1px">\nCloudFront attempted to establish a connection with the origin, but either the attempt failed or the origin closed the connection.\nWe can\'t connect to the server for this app or website at this time. There might be too much traffic or a configuration error. Try again later, or contact the app or website owner.\n<BR clear="all">\nIf you provide content to customers through CloudFront, you can find steps to troubleshoot and help prevent this error by reviewing the CloudFront documentation.\n<BR clear="all">\n<HR noshade size="1px">\n<PRE>\nGenerated by cloudfront (CloudFront)\nRequest ID: WWzX15y3kJbF0AZuWI90otaXKOYm34OpoNQSoFHu2SmN5bibxYw03A==\n</PRE>\n<ADDRESS>\n</ADDRESS>\n</BODY></HTML>'
```

```
Status: 502 Content: b'<html>\r\n<head><title>502 Bad Gateway</title></head>\r\n<body>\r\n<center><h1>502 Bad Gateway</h1></center>\r\n</body>\r\n</html>\r\n'
```

```
Status: 503 Content: b'<html>\r\n<head><title>503 Service Temporarily Unavailable</title></head>\r\n<body>\r\n<center><h1>503 Service Temporarily Unavailable</h1></center>\r\n</body>\r\n</html>\r\n'
```

This screenshot shows the last entries in the log of one of the tasks that was killed. Gunicorn killed the worker after a critical timeout and presumably the new worker could not respond to health checks.

Screen Shot 2020-09-30 at 2 38 50 PM

**Comment by jwalgran** [comment]
I realized that while I was running the streams I was trying to use the Django admin to delete some data and try and induce some changes that would require model updates. I visited https://staging.openapparel.org/admin/api/facilitylistitem/131527/change/ and it hung. We have an open issue (#1115) about this page is taking too long to load.

I started a single stream of post requests, attempted to visit the admin URL again, and induced 504 responses from CloudFront.

My understanding as of now is that this request exceeded the 120s gunicorn worker timeout, triggering the termination of the worker and staring a replacement worker which would then not be able to respond to health checks within the allowed time frame.

I am going to see if I can quickly resolve #1115

**Comment by jwalgran** [comment]
I was able to get the `GazetterCache` into a bad state by deleting a facility, which triggers exceptions inside of Dedupe when calling `unindex`. Deleting a facility is a  rare occurrence, and it cannot be done entirely through the admin UI as of now, but it does happen. 

Exceptions raised in Rollbar look like this

Screen Shot 2020-09-30 at 3 40 38 PM

Screen Shot 2020-09-30 at 6 06 27 PM

I looked at the differences between the version of dedupe we are running (1.9.4) and the latest version in the 1.x series (1.10.0) specifically looking for code in the stack traces that was changed, but did not see anything significant, just a change from a Python "future" method to native dict iteration.

https://github.com/dedupeio/dedupe/compare/v1.9.4...v1.10.0

I pushed a commit to fix up our custom watchman checker so that an exception raised by `GazetteerCache.get_latest` ends up as a 503.

After deploying this commit I saw some interesting behavior. When I POSTed then deleted an "Azavea" facility, I saw one of the instances of the service get into the bad state and eventually get replaced. This is good, and proves that my update to the health checker works.

Screen Shot 2020-09-30 at 5 42 04 PM

Unexpectedly only one instance ran into this unindex exception. The other instance remained healthy. You can see the flip flopping of the health check endpoint in this screencast



I have not yet been able to reproduce the crash on unindex in development.

While digging into dedupe code and docs, I did see that the company behind it were running a REST API connected to there web service. The documentation for the REST API has an "add" endpoint for adding records and a "match" endpoint for matching a single record but interestingly it does not have an unindex or remove endpoint.

https://apidocs.dedupe.io/en/latest/index.html
Screen Shot 2020-09-30 at 5 56 29 PM

This is pushing me toward the idea of catching exceptions in `unindex`, logging them without crashing, and updating the match logic to discard and matches for facilities that no longer exist, since the model will still "know" about the deleted facility.

**Comment by jwalgran** [comment]
While working on gracefully handling of deleted facilities I recalled that when facilities are merged through the dashboard moderation tools, we delete the "duplicate" facility. Deleting facilities outright is unusual, but merging is a common maintenance activity. This reinforces my belief that working around problems with deletes is the correct way to go.

**Comment by jwalgran** [comment]
036df92 is my first pass at reworking the `GazetteerCache` so that certain exceptions thrown out of the Dedupe `unindex` method are logged to Rollbar rather then allowed to crash and/or leave the model in an un-updatable state. To cope with that change we now filter out matches for facilities that may still be in the model, but have been removed from the database.

While testing that I was trying to use the http://localhost:6543/dashboard/mergefacilities tool and discovered a bug that was introduced when we started allowing API submissions with `create=false`. I addressed this in 0e562a9 by also cleaning up these items as part of the merge operation.

**Comment by jwalgran** [comment]
I went through these steps to test the new matching system on staging

- Started ./tools/postfacilitiescsv.py with a data file provided by our API partner
- Used the swagger UI to post these two new facilities and copy their OAR IDs
```
{
    "country": "US",
    "address": "990 Spring Garden, Philadelphia",
    "name": "Azavea"
}
```
```
{
    "country": "US",
    "address": "Philadelphia",
    "name": "Azavea Distinct Match"
}
```
- Used http://staging.openapparel.org/dashboard/mergefacilities to merge the second into the first, which effectively deleting the second
- POSTed the second facility again, which created a new facility record, as expected
```
{
    "country": "US",
    "address": "Philadelphia",
    "name": "Azavea Distinct Match"
}
```
- I saw 500 errors being returned from `./tools/postfacilitiescsv.py`. I looked one up one of the  failing items in the admin (https://staging.openapparel.org/admin/api/facilitylistitem/) and got the trace from the `processing_results`

```
    "trace": "Traceback (most recent call last):
  File \"/usr/local/src/api/views.py\", line 1132, in create
    match_results = match_item(country_code, name, address, item.id)
  File \"/usr/local/src/api/matching.py\", line 412, in match_item
    recall_weight=recall_weight)
  File \"/usr/local/src/api/matching.py\", line 289, in match_items
    gazetteer.threshold(messy, recall_weight=recall_weight)
  File \"/usr/local/lib/python3.7/site-packages/dedupe/api.py\", line 975, in threshold
    return self.thresholdBlocks(blocked_pairs, recall_weight)
  File \"/usr/local/lib/python3.7/site-packages/dedupe/api.py\", line 69, in thresholdBlocks
    candidate_records = itertools.chain.from_iterable(self._blockedPairs(blocks))
  File \"/usr/local/lib/python3.7/site-packages/dedupe/api.py\", line 406, in _blockedPairs
    block, blocks = core.peek(blocks)
  File \"/usr/local/lib/python3.7/site-packages/dedupe/core.py\", line 383, in peek
    record = next(records)
  File \"/usr/local/lib/python3.7/site-packages/dedupe/api.py\", line 863, in _blockData
    for each in self._blockGenerator(messy_data, self.blocked_records):
  File \"/usr/local/lib/python3.7/site-packages/dedupe/api.py\", line 427, in _blockGenerator
    for block_key, _ in block_keys:
  File \"/usr/local/lib/python3.7/site-packages/dedupe/blocking.py\", line 39, in __call__
    block_keys = predicate(instance, target=target)
  File \"/usr/local/lib/python3.7/site-packages/dedupe/predicates.py\", line 287, in __call__
    for predicate in self]
  File \"/usr/local/lib/python3.7/site-packages/dedupe/predicates.py\", line 287, in 
    for predicate in self]
  File \"/usr/local/lib/python3.7/site-packages/dedupe/predicates.py\", line 192, in __call__
    centers = self.index.search(doc, self.threshold)
  File \"/usr/local/lib/python3.7/site-packages/dedupe/tfidf.py\", line 36, in search
    in self._index.apply(query_list, threshold)]
  File \"/usr/local/lib/python3.7/site-packages/dedupe/canopy_index.py\", line 54, in apply
    docs = _wordinfo[wid]
KeyError: 1
```

My interpretation of this is that the unhandled exception within `unindex` left the model in a state where subsequent matches fail.

To attempt to deal with this I am going to remove the calls to `unindex` all together.

**Comment by jwalgran** [comment]
After removing the calls to `unindex` in 5515132 I was able to merge and delete facilities while running 4 `./tools/postfacilitiescsv.py` processes simultaneously. The tasks have not been killed and have been returning 2xx and 4xx responses as they should for about 20 minutes. I am comfortable having our API partner test again with this branch deployed to staging.

**Review by hectcastro — APPROVED** [review]
👍

File: docker-compose.yml [hunk 1]
```diff
@@ -39,7 +39,9 @@ services:
     command:
         - '-b :8081'
         - '--reload'
+        - '--timeout=60'
```
[the remark this answers is not in the document]
Comment (jwalgran) [reply docker-compose.yml:42]: Yep. Fixed. Thanks.

**Pull request merged**